File: 品味Java的21種設計模式/dp_cpp/iterator/example7/ArrayIteratorImpl.cpp
#include "ArrayIteratorImpl.h"

namespace cn
{
	namespace javass
	{
		namespace dp
		{
			namespace iterator
			{
				namespace example7
				{

					ArrayIteratorImpl::ArrayIteratorImpl(SalaryManager *aggregate)
					{
						InitializeInstanceFields();
						this->aggregate = aggregate;
					}

					void ArrayIteratorImpl::first()
					{
						index = 0;
					}

					void ArrayIteratorImpl::next()
					{
						if(index < this->aggregate->size())
						{
							index = index + 1;
						}
					}

					bool ArrayIteratorImpl::isDone()
					{
						if(index == this->aggregate->size())
						{
							return true;
						}
						return false;
					}

					object *ArrayIteratorImpl::currentItem()
					{
						return this->aggregate->get(index);
					}

					bool ArrayIteratorImpl::isFirst()
					{
						if(index==0)
						{
							return true;
						}
						return false;
					}

					void ArrayIteratorImpl::previous()
					{
						if(index > 0)
						{
							index = index - 1;
						}
					}
				}
			}
		}
	}
}

File: 品味Java的21種設計模式/dp_java/iterator/example7/ArrayIteratorImpl.java
package cn.javass.dp.iterator.example7;

/**
 * 用来实现访问数组的双向迭代接口
 */
public class ArrayIteratorImpl implements Iterator {
    /**
     * 用来存放被迭代的聚合对象
     */
    private SalaryManager aggregate = null;
    /**
     * 用来记录当前迭代到的位置索引
     * -1表示刚开始的时候，迭代器指向聚合对象第一个对象之前
     */
    private int index = -1;

    public ArrayIteratorImpl(SalaryManager aggregate) {
        this.aggregate = aggregate;
    }

    public void first() {
        index = 0;
    }
    public void next() {
        if (index < this.aggregate.size()) {
            index = index + 1;
        }
    }
    public boolean isDone() {
        if (index == this.aggregate.size()) {
            return true;
        }

        return false;
    }
    public Object currentItem() {
        return this.aggregate.get(index);
    }

    public boolean isFirst() {
        if (index == 0) {
            return true;
        }

        return false;
    }
    public void previous() {
        if (index > 0) {
            index = index - 1;
        }
    }
}
